Add MimbleWimble support

webwarrior-ws opened the pull request:
Implemented parsing and creation of MimbleWimble transactions, as well as Wallet type for keeping track of unspent coins. Added tests for implemented functionality.

webwarrior-ws committed:
MimbleWimble(Tests): implement MW types

Add types needed for MimbleWimble transactions and implement
serialization.

Made all MimbleWimble types implement ISerializable and
interface and Read static method. Reuse NBitcoin's
BitcoinStream for serialization and deserialization, but don't
use its IBitcoinSerializable interface because it mixes reading
and writing in one method.

Added new test project for MimbleWimble tests. Added test for
peg-in transaction.

Added property-based tests for serialization and
deserilaization.

Added test that deserilizes transaction generated and
serialized by modified litecoin test (see [1]).

Added tests for parsing peg-out transaction and MW to MW
(HogEx) transaction.

[1] https://github.com/litecoin-project/litecoin/blob/5ac781487cc9589131437b23c69829f04002b97e/src/libmw/test/tests/models/tx/Test_Transaction.cpp

webwarrior-ws committed:
MimbleWimble: started implementing MW tx creation

Add NBitcoin.Secp256k1 library because it has Schnorr
signature. As the library requires .netstandard 2.1,
make NLitecoin target it instead of 2.0.

webwarrior-ws committed:
MimbleWimble(Tests): implement tx builder

Implement transaction builder and add tests for it.

webwarrior-ws committed:
MimbleWimble(Tests): test key derivation

Ported test for key derivation and stealth address generation
using KeyChain.

Minor code style improvements.

webwarrior-ws committed:
MimbleWimble: fixes in Wallet and Tx builder

Fixes in Wallet and TransactionBuilder mostly related to
public/private spend keys.

webwarrior-ws committed:
MimbleWimble.Tests: added test for Wallet

Added integration test for Wallet.

webwarrior-ws committed:
Nlitecoin: update NBitcoin to 7.0.13

Update NBitcoin to version 7.0.13 because older NBitcoin
versions can't decode bech32 strings of length > 90, which is
needed for MimbleWimble address decoding.

webwarrior-ws committed:
MimbleWimble(Tests): implemented MW address decoding

Implemented decoding of MimbleWimble stealth addresses. Added
test.

webwarrior-ws committed:
Nlitecoin: bump version

Increase package's version to 0.2.0.

knocte commented:
I know you have squashed all commits into 1 in order to not have to curate your commit history to have all commits be green. But I prefer a not-all-green commit history for a PR (that I squash into one myself when merging) than just losing all those commits that you had @webwarrior-ws

webwarrior-ws commented:
> I know you have squashed all commits into 1 in order to not have to curate your commit history to have all commits be green. But I prefer a not-all-green commit history for a PR (that I squash into one myself when merging) than just losing all those commits that you had @webwarrior-ws

What do you propose? To have all the original commits in the PR?

knocte commented:
That's exactly what I said.

knocte commented:
i.e. ...**But I prefer a not-all-green commit history for a PR**...

webwarrior-ws committed:
MimbleWimble: implement readonly keychain

Implement "read-only" keychain, that allows for balance to be
checked without knowledge of wallet's seed (which would allow
for spending funds), only with private scan key and public
spend keys.

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+open Org.BouncyCastle.Math.EC
+open Org.BouncyCastle.Crypto.Digests
+
+let curve = ECNamedCurveTable.GetByName("secp256k1")
```
@webwarrior-ws nit: unneeded parens

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+
+type NBitcoin.Secp256k1.ECPrivKey with
+    member self.ToBytes() =
+        let bytes = Array.zeroCreate 32
```
@webwarrior-ws magic number

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+        let hasher = Sha256Digest()
+        hasher.BlockUpdate(key, 0, key.Length)
+        hasher.BlockUpdate(msgHash, 0, msgHash.Length)
+        let arr = Array.zeroCreate 32
```
@webwarrior-ws nit: magic number

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+        BigInteger.FromByteArrayUnsigned(arr).Mod(scalarOrder)
+
+    if k0 = BigInteger.Zero then
+        failwith "Failure. This happens only with negligible probability."
```
@webwarrior-ws WHAT???

webwarrior-ws replied:
It's extremely unlikely that sha256 hash of some value will be all zeros or exactly scalarOrder.

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+        failwith "Failure. This happens only with negligible probability."
+    
+    let keyScalar = BigInteger.FromByteArrayUnsigned key
+    assert(keyScalar < scalarOrder)
```
@webwarrior-ws please use Fsx's Misc.BetterAssert

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+    let k = if Jakobi R.AffineYCoord <> 1 then scalarOrder.Subtract k0 else k0
+    let e = 
+        let hasher = Sha256Digest()
+        hasher.BlockUpdate(R.AffineXCoord.GetEncoded(), 0, 32)
```
@webwarrior-ws magic number

knocte commented on src/NLitecoin/MimbleWimble/EC.fs:
```diff
@@ -0,0 +1,120 @@
+        hasher.BlockUpdate(R.AffineXCoord.GetEncoded(), 0, 32)
+        hasher.BlockUpdate(generatorG.Multiply(keyScalar).GetEncoded(true), 0, 33)
+        hasher.BlockUpdate(msgHash, 0, msgHash.Length)
+        let arr = Array.zeroCreate 32
```
@webwarrior-ws magic number

knocte commented on src/NLitecoin/MimbleWimble/TransactionBuilder.fs:
```diff
@@ -0,0 +1,368 @@
+    (pegouts: array<PegOutCoin>) 
+    : Kernel =
+    let featuresByte =
+        (if fee > 0L then KernelFeatures.FEE_FEATURE_BIT else enum 0) |||
```
@webwarrior-ws please use fantomas to get better formatting here, putting if then else in a single line is not good

knocte commented on src/NLitecoin/MimbleWimble/TransactionBuilder.fs:
```diff
@@ -0,0 +1,368 @@
+        (match peginAmount with 
+            | Some value when value > 0L -> KernelFeatures.PEGIN_FEATURE_BIT 
+            | _ -> enum 0) |||
+        (if pegouts.Length > 0 then KernelFeatures.PEGOUT_FEATURE_BIT else enum 0) |||
```
@webwarrior-ws same here

knocte commented on src/NLitecoin/MimbleWimble/TransactionBuilder.fs:
```diff
@@ -0,0 +1,368 @@
+                pegoutTotal
+                recipientTotal
+                fee)
+        raise (IncorrectBalanceException msg)
```
@webwarrior-ws nit: use <| to avoid parens

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+type CAmount = int64
+
+[<AutoOpen>]
+module Helpers =
```
@webwarrior-ws please respect our F# style guideines, all funcs (in this module) should be PascalCase, not camelCase

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+        sprintf "BigInt %s" (encoder.EncodeData bytes)
+
+    static member Read(stream: BitcoinStream) (numBytes: int) : BigInt =
+        assert(not stream.Serializing)
```
@webwarrior-ws same here wrt assert

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+
+    interface ISerializeable with
+        member self.Write(stream) =
+            assert(stream.Serializing)
```
@webwarrior-ws same here

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+    static member Read(stream: BitcoinStream) : OutputMessage =
+        assert(not stream.Serializing)
+
+        let featuresByte = ref 0uy
```
@webwarrior-ws why do you need refs here, micro-optimization?

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+    static member Size = 675
+
+    static member Read(stream: BitcoinStream) : RangeProof =
+        assert(not stream.Serializing)
```
@webwarrior-ws pls replace all asserts with BetterAssert

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+        }
+
+    interface ISerializeable with
+        member self.Write(stream) = 
```
@webwarrior-ws nit: unneeded parens

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+            write stream self.Signature
+
+    interface IComparable<Output> with
+        member self.CompareTo(other) =
```
@webwarrior-ws nit: unneeded parens

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+        }
+
+    interface ISerializeable with
+        member self.Write(stream) = 
```
@webwarrior-ws nit: unneeded parens

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+    member self.IsMine = self.AddressIndex <> Coin.UnknownIndex
+    member self.HasSpendKey = self.SpendKey.IsSome
+
+    static member Empty =
```
@webwarrior-ws can this be marked as private or internal?

webwarrior-ws replied:
No, it's used in tests. Unless I duplicate this method in tests.

knocte replied:
don't duplicate it

knocte commented on src/NLitecoin/MimbleWimble/Wallet.fs:
```diff
@@ -0,0 +1,377 @@
+                    let spendPubKey = (calculateSpendKey i).CreatePubKey()
+                    Collections.Generic.KeyValuePair(spendPubKey, i) })
+
+    new(seed: array<byte>) = KeyChain(seed, 100u)
```
@webwarrior-ws magic number

knocte commented on src/NLitecoin/MimbleWimble/Wallet.fs:
```diff
@@ -0,0 +1,377 @@
+
+    interface IKeyChain with
+        override self.GetStealthAddress(index) = self.GetStealthAddress(index)
+        override self.RewindOutput(output) = self.RewindOutput(output)
```
@webwarrior-ws nit: unneeded parens in 2 lines above

knocte commented on src/NLitecoin/MimbleWimble/Types.fs:
```diff
@@ -0,0 +1,789 @@
+    abstract Write: BitcoinStream -> unit
+    // no Read() method as it will be static method and can't be included in interface
+
+type CAmount = int64
```
@webwarrior-ws what does C stand for? please rename

webwarrior-ws committed:
MimbleWimble(Tests): addressed comments

Addressed code review comments. Added latest Fsdk package in
order to use BetterAssert function.

webwarrior-ws committed:
NLitecoin,MimbleWimble(Tests): handle HogEx Txs

Fixed transaction deserialization so that it now handles HogEx
transactions, which have mweb extension flag but don't contain
MW transaction. Added test for this case.

knocte commented:
@webwarrior-ws I guess this PR needs cleanup (e.g. squashing). But if you add LIP0006 work here, please add it as a separate commit.

webwarrior-ws committed:
MimbleWimble(Tests): implement range proofs

Implement construction of range proofs.

Added a new test project that does differential tests for
zero-knowledge proof components that uses Secp256k1-ZKP.Net
library (libsecp256k1 wrapper) as reference.

Use Assert.Inconclusive if function is missing in libsecp256k1
on current platform.

webwarrior-ws committed:
MimbleWimble(Tests): implement Wallet

Added Wallet type, which is responsible for key derivation,
rewinding of transaction outputs, and storage of rewound
outputs (coins), as well as creating transactions using our
keychain and spendable coins.

webwarrior-ws commented:
Squashed. CI for all commits is green now.
